### ISSUE TITLE
[Bug] Overlap scheduler silently corrupts tokens after multiple padded prefill batches

### ISSUE THREAD

**Issue by Yanko-7** [sgl-project/sglang-jax#1487]
### Checklist

- [x] 1. I have searched related issues but cannot get the expected help.
- [x] 2. The bug has not been fixed in the latest version.
- [x] 3. Please note that if the bug-related issue you submitted lacks corresponding environment info and a minimal reproducible demo, it will be challenging for us to reproduce and resolve the issue, reducing the likelihood of receiving feedback.
- [x] 4. If the issue you raised is not a bug but a question, please raise a discussion at https://github.com/sgl-project/sgl-jax/discussions/new/choose Otherwise, it will be closed.
- [x] 5. Please use English, otherwise it will be closed.

### Describe the bug

On current `main`, overlap scheduling can silently replace a request's pending
future token with a token produced by a later batch. Requests complete without
an exception, but their generated text differs from the overlap-off baseline.

This was originally observed with heterogeneous Qwen2.5-VL requests and is
related to https://github.com/QwenLM/Qwen3-VL/issues/759. The reproduction below
uses the text-only `Qwen/Qwen2.5-1.5B-Instruct`, demonstrating that this is not
VLM-specific.

The future-token ring is limited to `max_running_requests * 3`:

https://github.com/sgl-project/sglang-jax/blob/b859cf172e9410b87124236ea818fb389b0355fa/python/sgl_jax/srt/managers/tp_worker_overlap_thread.py#L48-L51

Every forward advances the counter by `len(model_worker_batch.seq_lens)`, which
is the padded batch width:

https://github.com/sgl-project/sglang-jax/blob/b859cf172e9410b87124236ea818fb389b0355fa/python/sgl_jax/srt/managers/tp_worker_overlap_thread.py#L297-L306

With `max_running_requests=32` and batch padding 32, the effective 96-slot ring
repeats after three forwards:

```text
forward 1: slots  1..32
forward 2: slots 33..64
forward 3: slots 65..96
forward 4: slots  1..32  # overwrites forward 1
```

If a completed prefill has not decoded yet, its negative placeholder resolves
to the newer token. Expanding the ring only delays the same lifetime problem;
the relay needs request ownership or generation/lifetime tracking.

### Reproduction

Tested at:

```text
b859cf172e9410b87124236ea818fb389b0355fa
```

### 1. Install

```bash
git clone https://github.com/sgl-project/sglang-jax.git
cd sglang-jax
git checkout b859cf172e9410b87124236ea818fb389b0355fa
cd python
uv sync --extra tpu
source .venv/bin/activate
cd ..
```

### 2. Save the server launcher

```bash
cat >/tmp/run_overlap_repro_server.sh <<'EOF'
#!/usr/bin/env bash
set -euo pipefail

overlap_args=()
if [[ "${1:-on}" == "off" ]]; then
  overlap_args+=(--disable-overlap-schedule)
fi

python -u -m sgl_jax.launch_server \
  --model-path Qwen/Qwen2.5-1.5B-Instruct \
  --trust-remote-code \
  --dtype bfloat16 \
  --mem-fraction-static 0.8 \
  --max-running-requests 32 \
  --max-prefill-tokens 512 \
  --chunked-prefill-size 128 \
  --page-size 64 \
  --device tpu \
  --tp-size 4 \
  --skip-server-warmup \
  --attention-backend fa \
  --max-seq-len 2048 \
  --precompile-token-paddings 512 \
  --precompile-bs-paddings 32 \
  --log-level info \
  --port 30000 \
  "${overlap_args[@]}"
EOF
chmod +x /tmp/run_overlap_repro_server.sh
```

### 3. Save the dependency-free client

```python
#!/usr/bin/env python3
import argparse
import json
import sys
import urllib.request
from concurrent.futures import ThreadPoolExecutor, as_completed


def send_request(case_index):
    marker = f"CASE-{case_index:02d}::OK"
    filler = " ".join(["context"] * (260 + case_index * 17))
    payload = {
        "model": "Qwen/Qwen2.5-1.5B-Instruct",
        "messages": [
            {
                "role": "system",
                "content": "Follow the final instruction exactly and do not explain.",
            },
            {
                "role": "user",
                "content": (
                    f"Background text follows: {filler}\n"
                    f"Final instruction: output exactly {marker}"
                ),
            },
        ],
        "temperature": 0,
        "max_tokens": 16,
    }
    request = urllib.request.Request(
        "http://127.0.0.1:30000/v1/chat/completions",
        data=json.dumps(payload).encode(),
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    with urllib.request.urlopen(request, timeout=900) as response:
        result = json.load(response)
    return result["choices"][0]["message"]["content"]


parser = argparse.ArgumentParser()
parser.add_argument("--baseline")
args = parser.parse_args()

outputs = {}
with ThreadPoolExecutor(max_workers=32) as executor:
    futures = {executor.submit(send_request, i): i for i in range(32)}
    for future in as_completed(futures):
        outputs[f"{futures[future]:02d}"] = future.result()
outputs = dict(sorted(outputs.items()))

if not args.baseline:
    json.dump(outputs, sys.stdout, indent=2)
    print()
else:
    with open(args.baseline) as file:
        baseline = json.load(file)
    differences = {
        key: {"overlap_off": baseline[key], "overlap_on": outputs[key]}
        for key in baseline
        if baseline[key] != outputs[key]
    }
    print(json.dumps(differences, indent=2))
    if differences:
        print(f"REPRODUCED: {len(differences)}/32 outputs differ", file=sys.stderr)
        raise SystemExit(1)
    print("NOT REPRODUCED: outputs are identical", file=sys.stderr)
```

Save it as `/tmp/repro_overlap.py`.

### 4. Capture the overlap-off baseline

In terminal 1:

```bash
/tmp/run_overlap_repro_server.sh off
```

After `/health` is ready, in terminal 2:

```bash
python /tmp/repro_overlap.py >/tmp/overlap_off.json
```

Stop the server, then restart it with overlap enabled:

```bash
/tmp/run_overlap_repro_server.sh on
```

Run the comparison:

```bash
python /tmp/repro_overlap.py --baseline /tmp/overlap_off.json
```

The exact script above reproduced the bug:

```text
REPRODUCED: 3/32 outputs differ

"03": {
  "overlap_off": "CASE-03::OK",
  "overlap_on": "CASE context context context context context ..."
}
```

Three fresh overlap-on server runs reproduced 3, 5, and 9 corrupted responses
out of 32. The affected request IDs vary with request arrival timing. The
server logged 26–27 prefill batches at request startup and no exception.

Expected: greedy outputs are identical with overlap on and off.

Actual: overlap-on silently changes some generated token sequences.

## Environment

Output from `python3 -m sgl_jax.check_env`:

```text
Python: 3.12.13 (main, Jul 18 2026, 17:02:19) [Clang 22.1.3 ]
sglang-jax: 0.0.0.dev1+gb859cf172
jax: 0.10.2
jaxlib: 0.10.2
triton: Module Not Found
transformers: 4.57.6
numpy: 2.2.6
aiohttp: 3.14.3
fastapi: 0.116.2
huggingface_hub: 0.34.6
modelscope: 1.28.2
orjson: 3.11.9
packaging: 26.2
psutil: 7.0.0
pydantic: 2.11.10
python-multipart: 0.0.32
pyzmq: 27.0.2
uvicorn: 0.35.0
uvloop: 0.21.0
openai: 2.48.0
tiktoken: 0.13.0
[TPU v6 lite-0]: TPU_0(process=0,(0,0,0,0))
[TPU v6 lite-1]: TPU_1(process=0,(1,0,0,0))
[TPU v6 lite-2]: TPU_2(process=0,(0,1,0,0))
[TPU v6 lite-3]: TPU_3(process=0,(1,1,0,0))
ulimit soft: 1048576
```

**Comment by real3stone** [comment]
Hi @Yanko-7 — great writeup, the buffer-sizing math made the root cause easy to follow. Are you planning to take this one yourself, or is it open for contribution?

Context: I just submitted [#1502](https://github.com/sgl-project/sglang-jax/pull/1502) fixing the batched-decode mask bug in the native attention backend ([#1297](https://github.com/sgl-project/sglang-jax/issues/1297)) — same family as this one: padding leaking into bookkeeping and silently corrupting results. I'd like to take on the FutureMap redesign next.

If it's open: I'd post a design sketch here within 2–3 days for feedback, then aim for a review-ready PR in about two weeks. If it's more urgent than that, a stopgap ring-enlargement could land right away to buy time — or feel free to just take it.

**Comment by Yanko-7** [comment]
> Hi [@Yanko-7](https://github.com/Yanko-7) — great writeup, the buffer-sizing math made the root cause easy to follow. Are you planning to take this one yourself, or is it open for contribution?
> 
> Context: I just submitted [#1502](https://github.com/sgl-project/sglang-jax/pull/1502) fixing the batched-decode mask bug in the native attention backend ([#1297](https://github.com/sgl-project/sglang-jax/issues/1297)) — same family as this one: padding leaking into bookkeeping and silently corrupting results. I'd like to take on the FutureMap redesign next.
> 
> If it's open: I'd post a design sketch here within 2–3 days for feedback, then aim for a review-ready PR in about two weeks. If it's more urgent than that, a stopgap ring-enlargement could land right away to buy time — or feel free to just take it.

Hi! Thanks for the feedback.

I'm focus on other tasks right now, so please feel free to drive this!
For reference, I previously put together a temporary workaround—it showed a modest (~1.5%) gain in a few scenarios, but isn't fully tested yet. Linking it here in case it's helpful for your redesign:https://github.com/primatrix/sglang-jax/commit/a5c6395b5bc64f28d5c98fd49a30ee1fab859a4c
Looking forward to the design sketch!

**Comment by real3stone** [comment]
## Design sketch: request-owned future-token relay

Thanks @Yanko-7 — taking this one. Sketch below; feedback welcome, especially on the fused-path question at the end.

### Why ownership, not a bigger ring

The ring's slot identity is temporal — "the k-th allocation" — while a request's placeholder lifetime is unbounded: under prefill-priority scheduling, a prefill-complete request waits for the next decode batch, and nothing bounds how many forwards that takes. Any ring size only resizes the overwrite window, and because the counter advances by the padded batch width even when most rows are padding, the window collapses to ~3 forwards however large the ring is.

Upstream PyTorch sglang is strong precedent for the direction: it patched ring sizing twice for this same bug class (sgl-project/sglang#13713, sgl-project/sglang#13916 — chunked prefill overrunning the ring), then abandoned the ring entirely in sgl-project/sglang#25862, moving to a relay indexed by `req_pool_idx` — slot identity bound to request lifetime, with a reserved row for padded batches.

### Design

Key the relay by the request's own pool slot, so a pending token cannot be clobbered by any other request for as long as the owner holds its `req_pool_idx`.

- `future_token_ids_map: int32[req_pool_size + 1]`, slot 0 reserved for padded rows. Sized from the worker's own `req_to_token_pool` — not `max_running_requests` — since per-worker pools can differ under PD.
- Placeholders become `-(req_pool_idx + 1)`; `resolve_future_token_ids` is unchanged.
- `set_future_token_ids` scatters by `req_pool_indices` instead of slicing at a counter; `future_token_ids_ct` / `future_token_ids_limit` are deleted outright.
- The scatter index is masked: `write_idx = jnp.where(req_pool_indices >= 0, req_pool_indices + 1, 0)`. Padded rows carry `req_pool_idx == -1` today, so `-1 + 1 = 0` would land on the reserved slot anyway — but `jnp`'s `.at[]` wraps negative indices NumPy-style, so any value ≤ −2 would silently write into a live slot at the tail of the map. The mask makes correctness independent of the padding convention rather than dependent on it.
- Correctness no longer rests on a lifetime bound but on two orderings, which I'll document in code comments since the design breaks silently if either changes:
  1. **queue FIFO** — the per-client forward thread consumes `input_queue` in order, so batch N's resolve always executes after batch N−1's scatter;
  2. **single-writer scheduler** — every pool free/re-alloc and every queue submission happens on the one scheduler thread, so when a slot is recycled, the new owner enqueues strictly after all of the old owner's in-flight batches.

Scope note: the relay logic exists in two places — the `managers/utils.py` helpers, and inlined in the fused jit (`jitted_run_and_sample`) that Pathways-PD decode uses when `SGLANG_PD_FUSE_SAMPLE=1`. Both get the change in this PR, along with the precompile stubs that pin the map's shape: fixing only the helpers would leave the fused path on the old ring, i.e. still corrupting, just in a less-travelled deployment. To stop the two copies drifting apart again, I'd like to factor the index arithmetic — placeholder encoding and the masked scatter index — into one shared helper that both call, leaving only the sharding plumbing path-specific.

### Test plan

- **T1 — unit (CPU)**: the relay contract on a 1-device mesh — a pending placeholder must resolve to its owner's token no matter how many padded forwards intervene. Should be red on current main (ring wraps after ~3 forwards), green with the fix. Extends `python/sgl_jax/test/test_tp_worker_overlap_thread.py` (already in the `unit-test-cpu` suite).
- **T2 — scheduler-level (CPU)**: `Scheduler.__new__`-style harness (same pattern as `test_scheduler_chunked_ownership.py`) driving prefill-complete → 3+ padded prefill batches → first decode, asserting the resolved token belongs to the owning request. Locks the host-side placeholder lifecycle against future refactors.
- **T3 — e2e regression (TPU)**: overlap on vs off greedy token-level equality on a small model with `max_running_requests=8` (ring wraps every 3 forwards) and a small `chunked_prefill_size` to keep decode starved — the issue's repro tuned toward deterministic triggering. Registered in `run_suite.py`. Token-level with deterministic sampling, to avoid the bf16 bit-exactness flakiness that got `test_engine_determine_generation` disabled.

### Plan

I'll start on the tests and implementation now, targeting a review-ready PR in ~2 weeks as discussed. Two asks: (1) any objection to keying the relay by `req_pool_idx` rather than a ring counter? (2) for the fused path, I plan to cover the relay semantics via the shared helper above plus a fake-device CPU harness in the style of `test/srt/disaggregation/test_pathways_pd.py` — I can't exercise a real Pathways deployment, so if that leaves a gap you'd want closed before merge, tell me now and I'll plan for it. (Incidentally, `test_pathways_pd.py` doesn't appear in `run_suite.py`, so it looks like it never runs in CI — happy to register it separately if that's just an oversight.)

**Comment by real3stone** [comment]
@Yanko-7 — heads-up: [#1504](https://github.com/sgl-project/sglang-jax/pull/1504) (opened by @zorrofox a few hours before my sketch above) fixes this exact root cause — padded-bs cursor advance wrapping the ring — with the same per-request-slot design, plus multi-host accuracy A/B I couldn't have produced. It references [#1317](https://github.com/real3stone/RogerPlan/issues/1317) rather than this issue, so it didn't show up here.

I've moved my effort to reviewing it rather than duplicating work. Suggest cross-referencing the two so this issue closes with [#1504](https://github.com/sgl-project/sglang-jax/pull/1504). Thanks for the handoff — the writeup made the root cause easy to follow, and the design sketch was worth doing regardless.

**Comment by Yanko-7** [comment]
This issue has been resolved. https://github.com/sgl-project/sglang-jax/pull/1504
